src/metabase/upload.clj

```diff
@@ -117,9 +117,26 @@
   [n values]
   (first (partition n n (repeat nil) values)))
 
+(defn- normalize-column-name
+  [[raw-name index]]
+  (let [normalized (u/slugify (str/trim raw-name))]
+    (if-not (str/blank? normalized)
+      normalized
+      (format "unnamed_column_%s" (inc index)))))
+
+(defn- deduplicate
+  "Add `new-name` to the vector (must be a vector so that `conj` works!) of `names-so-far`, adding a unique suffix if
+  necessary."
+  [names-so-far new-name]
+  (if (some #(= % new-name) names-so-far)
+    (if-let [dupe-number (second (re-matches #".*_(\d+)" new-name))]
+      (recur names-so-far (str/replace new-name #"_\d+$" (str "_" (inc (parse-long dupe-number)))))
+      (recur names-so-far (str new-name "_1")))
+    (conj names-so-far new-name)))
```

Comment:
Also, this will change the names of columns that are legitimately named `some_column_1`, which is not what we want. Either use `loop` and have an accumulator like `name-counts` keep track of how many occurrances of new-name there have been, or count how many dupes there are with `(count (filter #(= % new-name) names-so-far))`

Reply:
Recursion is necessary to handle admittedly-pathological cases:


```
(defn- cal-deduplicate
  "Add `new-name` to the vector (must be a vector so that `conj` works!) of `names-so-far`, adding a unique suffix if
  necessary."
  [names-so-far new-name]
  (let [new-name (if (some #(= % new-name) names-so-far)
                   (if-let [dupe-number (second (re-matches #".*_(\d+)" new-name))]
                     (str/replace new-name #"_\d+$" (str "_" (inc (parse-long dupe-number))))
                     (str new-name "_1"))
                   new-name)]
    (conj names-so-far new-name)))

(reduce cal-deduplicate [] ["a" "a" "a" "a_2"])
;; => ["a" "a_1" "a_1" "a_2"]

(reduce original-deduplicate [] ["a" "a" "a" "a_2"])
;; => ["a" "a_1" "a_2" "a_3"]
```

That said, while the `original-deduplicate` result above is _better_, it exemplifies your second point of clobbering columns that legitimately have a trailing number. I'm…working on it.

Reply:
True I understand now. If it sees an "a" and there is "a", "a_1", "a_2", etc you need to iterate until "a_N" is found. But yes, seems like you wouldn't need a recursive solution if you had an accumulator like `name-counts` though, as I think you figured out.